File: publish-checklist.py
#!/usr/bin/env python3
"""
Pre-publishing validation script for Alexa Discogs Skill
"""

import json
import os
import requests
from urllib.parse import urlparse

def validate_skill_config():
    """Validate skill configuration for publishing"""
    print("🎵 Validating Alexa Discogs Skill for Publishing...")
    print("=" * 50)
    
    issues = []
    warnings = []
    
    # Check if skill.json exists
    if not os.path.exists('skill-package/skill.json'):
        issues.append("❌ skill.json not found")
        return issues, warnings
    
    # Load skill configuration
    with open('skill-package/skill.json', 'r') as f:
        skill_config = json.load(f)
    
    manifest = skill_config.get('manifest', {})
    pub_info = manifest.get('publishingInformation', {})
    privacy = manifest.get('privacyAndCompliance', {})
    
    # Check publishing information
    locales = pub_info.get('locales', {})
    en_us = locales.get('en-US', {})
    
    # Validate required fields
    required_fields = ['name', 'summary', 'description', 'examplePhrases', 'keywords']
    for field in required_fields:
        if not en_us.get(field):
            issues.append(f"❌ Missing required field: {field}")
    
    # Check icons
    large_icon = en_us.get('largeIconUri', '')
    small_icon = en_us.get('smallIconUri', '')
    
    if 'example.com' in large_icon or not large_icon:
        issues.append("❌ Large icon URL is placeholder or missing")
    if 'example.com' in small_icon or not small_icon:
        issues.append("❌ Small icon URL is placeholder or missing")
    
    # Check privacy policy and terms
    privacy_locales = privacy.get('locales', {}).get('en-US', {})
    privacy_url = privacy_locales.get('privacyPolicyUrl', '')
    terms_url = privacy_locales.get('termsOfUseUrl', '')
    
    if 'example.com' in privacy_url or not privacy_url:
        issues.append("❌ Privacy policy URL is placeholder or missing")
    if 'example.com' in terms_url or not terms_url:
        issues.append("❌ Terms of use URL is placeholder or missing")
    
    # Check interaction model
    if not os.path.exists('skill-package/interactionModels/custom/en-US.json'):
        issues.append("❌ Interaction model not found")
    else:
        with open('skill-package/interactionModels/custom/en-US.json', 'r') as f:
            interaction_model = json.load(f)
        
        intents = interaction_model.get('interactionModel', {}).get('languageModel', {}).get('intents', [])
        custom_intents = [i for i in intents if not i['name'].startswith('AMAZON.')]
        
        if len(custom_intents) < 3:
            warnings.append("⚠️  Consider adding more custom intents for better functionality")
    
    # Check source files
    required_files = [
        'lambda_function.py',
        'requirements.txt',
        'src/handlers/basic_handlers.py',
        'src/handlers/list_releases_handler.py',
        'src/handlers/search_release_handler.py',
        'src/handlers/random_release_handler.py',
        'src/utils/discogs_utils.py'
    ]
    
    for file_path in required_files:
        if not os.path.exists(file_path):
            issues.append(f"❌ Missing required file: {file_path}")
    
    # Validate URLs (basic check)
    urls_to_check = []
    if large_icon and 'example.com' not in large_icon:
        urls_to_check.append(('Large Icon', large_icon))
    if small_icon and 'example.com' not in small_icon:
        urls_to_check.append(('Small Icon', small_icon))
    if privacy_url and 'example.com' not in privacy_url:
        urls_to_check.append(('Privacy Policy', privacy_url))
    if terms_url and 'example.com' not in terms_url:
        urls_to_check.append(('Terms of Use', terms_url))
    
    for name, url in urls_to_check:
        try:
            parsed = urlparse(url)
            if parsed.scheme != 'https':
                issues.append(f"❌ {name} URL must use HTTPS: {url}")
        except Exception:
            issues.append(f"❌ Invalid {name} URL: {url}")
    
    return issues, warnings

def print_results(issues, warnings):
    """Print validation results"""
    print("\n📋 VALIDATION RESULTS")
    print("=" * 50)
    
    if issues:
        print("🚨 CRITICAL ISSUES (Must fix before publishing):")
        for issue in issues:
            print(f"  {issue}")
    
    if warnings:
        print("\n⚠️  WARNINGS (Recommended to fix):")
        for warning in warnings:
            print(f"  {warning}")
    
    if not issues and not warnings:
        print("✅ All checks passed! Your skill is ready for publishing.")
    elif not issues:
        print("✅ No critical issues found. Address warnings if possible.")
    else:
        print(f"\n❌ Found {len(issues)} critical issues that must be fixed before publishing.")
    
    print("\n📚 Next Steps:")
    if issues:
        print("1. Fix all critical issues listed above")
        print("2. Re-run this validation script")
        print("3. Deploy to AWS Lambda")
        print("4. Test thoroughly in Alexa Developer Console")
        print("5. Submit for certification")
    else:
        print("1. Deploy to AWS Lambda using: python deploy.py")
        print("2. Test thoroughly in Alexa Developer Console")
        print("3. Submit for certification")
    
    print("\n📖 See PUBLISHING_GUIDE.md for detailed instructions")

def main():
    """Main validation function"""
    issues, warnings = validate_skill_config()
    print_results(issues, warnings)
    
    return len(issues) == 0

if __name__ == "__main__":
    success = main()
    exit(0 if success else 1)

File: publish-checklist.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { URL } = require('url');

function validateSkillConfig() {
    console.log('🎵 Validating Alexa Discogs Skill for Publishing (Node.js)...');
    console.log('='.repeat(50));
    
    const issues = [];
    const warnings = [];
    
    // Check if skill.json exists
    const skillJsonPath = path.join(__dirname, 'skill-package', 'skill.json');
    if (!fs.existsSync(skillJsonPath)) {
        issues.push('❌ skill.json not found');
        return { issues, warnings };
    }
    
    // Load skill configuration
    const skillConfig = JSON.parse(fs.readFileSync(skillJsonPath, 'utf8'));
    const manifest = skillConfig.manifest || {};
    const pubInfo = manifest.publishingInformation || {};
    const privacy = manifest.privacyAndCompliance || {};
    
    // Check publishing information
    const locales = pubInfo.locales || {};
    const enUs = locales['en-US'] || {};
    
    // Validate required fields
    const requiredFields = ['name', 'summary', 'description', 'examplePhrases', 'keywords'];
    requiredFields.forEach(field => {
        if (!enUs[field]) {
            issues.push(`❌ Missing required field: ${field}`);
        }
    });
    
    // Check icons
    const largeIcon = enUs.largeIconUri || '';
    const smallIcon = enUs.smallIconUri || '';
    
    if (largeIcon.includes('example.com') || !largeIcon) {
        issues.push('❌ Large icon URL is placeholder or missing');
    }
    if (smallIcon.includes('example.com') || !smallIcon) {
        issues.push('❌ Small icon URL is placeholder or missing');
    }
    
    // Check privacy policy and terms
    const privacyLocales = (privacy.locales || {})['en-US'] || {};
    const privacyUrl = privacyLocales.privacyPolicyUrl || '';
    const termsUrl = privacyLocales.termsOfUseUrl || '';
    
    if (privacyUrl.includes('example.com') || !privacyUrl) {
        issues.push('❌ Privacy policy URL is placeholder or missing');
    }
    if (termsUrl.includes('example.com') || !termsUrl) {
        issues.push('❌ Terms of use URL is placeholder or missing');
    }
    
    // Check interaction model
    const interactionModelPath = path.join(__dirname, 'skill-package', 'interactionModels', 'custom', 'en-US.json');
    if (!fs.existsSync(interactionModelPath)) {
        issues.push('❌ Interaction model not found');
    } else {
        const interactionModel = JSON.parse(fs.readFileSync(interactionModelPath, 'utf8'));
        const intents = (interactionModel.interactionModel?.languageModel?.intents) || [];
        const customIntents = intents.filter(i => !i.name.startsWith('AMAZON.'));
        
        if (customIntents.length < 3) {
            warnings.push('⚠️  Consider adding more custom intents for better functionality');
        }
    }
    
    // Check Node.js source files
    const requiredFiles = [
        'index.js',
        'package.json',
        'src/handlers/launchHandler.js',
        'src/handlers/listReleasesHandler.js',
        'src/handlers/searchReleaseHandler.js',
        'src/handlers/randomReleaseHandler.js',
        'src/utils/discogsUtils.js'
    ];
    
    requiredFiles.forEach(filePath => {
        if (!fs.existsSync(path.join(__dirname, filePath))) {
            issues.push(`❌ Missing required file: ${filePath}`);
        }
    });
    
    // Validate URLs (basic check)
    const urlsToCheck = [];
    if (largeIcon && !largeIcon.includes('example.com')) {
        urlsToCheck.push(['Large Icon', largeIcon]);
    }
    if (smallIcon && !smallIcon.includes('example.com')) {
        urlsToCheck.push(['Small Icon', smallIcon]);
    }
    if (privacyUrl && !privacyUrl.includes('example.com')) {
        urlsToCheck.push(['Privacy Policy', privacyUrl]);
    }
    if (termsUrl && !termsUrl.includes('example.com')) {
        urlsToCheck.push(['Terms of Use', termsUrl]);
    }
    
    urlsToCheck.forEach(([name, url]) => {
        try {
            const parsed = new URL(url);
            if (parsed.protocol !== 'https:') {
                issues.push(`❌ ${name} URL must use HTTPS: ${url}`);
            }
        } catch (error) {
            issues.push(`❌ Invalid ${name} URL: ${url}`);
        }
    });
    
    // Check package.json
    if (fs.existsSync('package.json')) {
        const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
        const requiredDeps = ['ask-sdk-core', 'ask-sdk-model', 'axios'];
        
        requiredDeps.forEach(dep => {
            if (!packageJson.dependencies || !packageJson.dependencies[dep]) {
                issues.push(`❌ Missing required dependency: ${dep}`);
            }
        });
        
        if (!packageJson.engines || !packageJson.engines.node) {
            warnings.push('⚠️  Consider specifying Node.js version in package.json engines field');
        }
    }
    
    return { issues, warnings };
}

function printResults(issues, warnings) {
    console.log('\n📋 VALIDATION RESULTS');
    console.log('='.repeat(50));
    
    if (issues.length > 0) {
        console.log('🚨 CRITICAL ISSUES (Must fix before publishing):');
        issues.forEach(issue => console.log(`  ${issue}`));
    }
    
    if (warnings.length > 0) {
        console.log('\n⚠️  WARNINGS (Recommended to fix):');
        warnings.forEach(warning => console.log(`  ${warning}`));
    }
    
    if (issues.length === 0 && warnings.length === 0) {
        console.log('✅ All checks passed! Your skill is ready for publishing.');
    } else if (issues.length === 0) {
        console.log('✅ No critical issues found. Address warnings if possible.');
    } else {
        console.log(`\n❌ Found ${issues.length} critical issues that must be fixed before publishing.`);
    }
    
    console.log('\n📚 Next Steps:');
    if (issues.length > 0) {
        console.log('1. Fix all critical issues listed above');
        console.log('2. Re-run this validation script');
        console.log('3. Install dependencies: npm install');
        console.log('4. Deploy to AWS Lambda: npm run deploy');
        console.log('5. Test thoroughly in Alexa Developer Console');
        console.log('6. Submit for certification');
    } else {
        console.log('1. Install dependencies: npm install');
        console.log('2. Deploy to AWS Lambda: npm run deploy');
        console.log('3. Test thoroughly in Alexa Developer Console');
        console.log('4. Submit for certification');
    }
    
    console.log('\n📖 See PUBLISHING_GUIDE.md for detailed instructions');
}

function main() {
    const { issues, warnings } = validateSkillConfig();
    printResults(issues, warnings);
    
    return issues.length === 0;
}

if (require.main === module) {
    const success = main();
    process.exit(success ? 0 : 1);
}

module.exports = { validateSkillConfig, printResults };